core/src/main/java/org/infinispan/container/offheap/OffHeapDataContainer.java

```diff
@@ -286,7 +293,7 @@ public boolean containsKey(Object k) {
 
          while (address != 0) {
             long nextAddress = offHeapEntryFactory.getNext(address);
-            if (offHeapEntryFactory.equalsKey(address, wba)) {
+            if (offHeapEntryFactory.equalsKey(address, wba) && !offHeapEntryFactory.isExpired(address)) {
```

Comment:
We should be able to short circuit this as we know that if the key equals then another can't equal.